src/mip/mip_logger.h

```diff
@@ -0,0 +1,200 @@
+/// and allows users to override the logging functions
+///
+///@{
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Logging level enum
+///
+typedef enum mip_logger_level
+{
+    MIP_LOGGER_LEVEL_OFF   = 0,  ///< Signifies that the log is turned off
+    MIP_LOGGER_LEVEL_FATAL = 1,  ///< Fatal logs are logged when an unrecoverable error occurs
+    MIP_LOGGER_LEVEL_ERROR = 2,  ///< Error logs are logged when an error occurs
+    MIP_LOGGER_LEVEL_WARN  = 3,  ///< Warning logs are logged when something concerning happens that may or not be a mistake
+    MIP_LOGGER_LEVEL_INFO  = 4,  ///< Info logs are logged when some general info needs to be conveyed to the user
+    MIP_LOGGER_LEVEL_DEBUG = 5,  ///< Debug logs are logged for debug purposes.
+    MIP_LOGGER_LEVEL_TRACE = 6,  ///< Trace logs are logged in similar cases to debug logs but can be logged in tight loops
+} mip_logger_level;
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Callback function typedef for custom logging behavior.
+///
+///@param context Context of what called this function. Could be a MIP device, serial connection, etc.
+///@param level   The log level that this log should be logged at
+///@param fmt     printf style format string
+///@param ...     Variadic args used to populate the fmt string
+///
+typedef void (*mip_logger_callback)(const void* context, void* user, mip_logger_level level, const char* fmt, ...);
+
+extern mip_logger_callback _mip_logger_callback;
+extern mip_logger_level _mip_logger_level;
+extern void* _mip_logger_user_data;
+
+void mip_logger_init(mip_logger_callback callback, mip_logger_level level, void* user);
+mip_logger_callback mip_logger_get_callback();
+mip_logger_level mip_logger_get_level();
+void* mip_logger_get_user_data();
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK. Prefer specific
+///       log level functions like mip_logger_info, etc. when possible.
+///@copydetails mip::C::mip_logger_callback
+#ifdef MIP_ENABLE_LOGGING
+#define mip_logger_log(context, level, fmt, ...) \
+    if (mip_logger_get_callback() != NULL && mip_logger_get_level() >= level) \
+        mip_logger_get_callback()(context, mip_logger_get_user_data(), level, fmt, __VA_ARGS__)
+#else
+#define mip_logger_log(context, level, fmt, ...) do {} while (1)
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+///@brief Helper macro used to log data inside the MIP SDK at fatal level
+///
+///@param context Context of what called this function. Could be a MIP device, serial connection, etc.
+///@param fmt     printf style format string
+///@param ...     Variadic args used to populate the fmt string
+///
+#define mip_logger_fatal(context, fmt, ...) mip_logger_log(context, MIP_LOGGER_LEVEL_FATAL, fmt, __VA_ARGS__)
```

Comment:
These should be all caps because they are macros. Also like the enum above let's shorten the name to MIP_LOG_XXX.